### PR TITLE
Prevent users from accidentally overwriting their source code

### LINKED ISSUE #4255: Error if package#source == package#main or package#module
# 🐛 bug report

Parcel 2 shouldn't overwrite the source in these situations:

```json
{
	"module": "src/index.js",
	"source": "src/index.js"
}
```

`parcel build .`

---

```json
{
	"main": "src/index.js",
}
```

`parcel build src/index.js`

### PR DESCRIPTION
**Pull request by JessieFrance**
# ↪️ Pull Request

This PR is for [#4255](https://github.com/parcel-bundler/parcel/issues/4255), and aims to prevent users from accidentally overwriting their own source code.

Users accidentally overwrite their own code when they configure their package.json main, module, browser, or distDir fields to be the same as the source entry file. This happened earlier in duplicates [#5281](https://github.com/parcel-bundler/parcel/issues/5281), and [#5041](https://github.com/parcel-bundler/parcel/issues/5041), and possibly also in [#5745](https://github.com/parcel-bundler/parcel/issues/5745).

This code patch would catch these previous cases, and would throw a diagnostic error with hints for package.json configurations, with a link to relevant Parcel V2 documentation.

## 💻 Examples
- Similar to the examples @mischnic gave in the [#4255](https://github.com/parcel-bundler/parcel/issues/4255) bug report, typing ```parcel build .``` in the terminal will overwrite src/index.js with this library mode package.json configuration:
```
{
	"main": "src/index.js",
	"source": "src/index.js"
}
```
In the above, replacing "main" with either "module" or "browser" will also overwrite src/index.js.

- Likewise, ```yarn run build``` with the custom target package.json configuration below will also overwrite source code:
```
{
  "targets": {
    "app": {
      "distDir": "src/index.html"
    }
  },
  "scripts": {
    "build": "parcel build src/index.html"
  }
}
```

## 🚨 Test instructions

Please see the proposed integration tests and run with
```yarn test:integration -g 'overwrite an entry'```

## ✔️ PR Todo

- [x] Added/updated ~~unit~~ integration tests for this change
- [x] Filled out test instructions (In case there aren't any unit tests)
- [x] Included links to related issues/PRs

Thank you for reviewing this PR. It's been fun working on this bug, and I'd love to hear your feedback.

Fixes #4255

### CODE REVIEW

**Comment by mischnic** [comment]
That check should rather be performed in here: https://github.com/parcel-bundler/parcel/blob/806cc65e0e430b1120cdcf1676e2f58738326d30/packages/core/core/src/requests/TargetRequest.js#L87-L108

(It would still execute in the same order as your current version, but its cleaner to put it into the TargetRequest itself)

**Comment by JessieFrance** [comment]
Thank you @mischnic for the helpful feedback on this. I moved the check over to the TargetRequest where you suggested (and did a rebase in the process to get rid of the last fix).

**Comment by JessieFrance** [comment]
@mischnic -- Sorry to disturb you, but I noticed that my integration test scenarios fail on the Windows CI pipeline, succeed on Linux, and possibly succeed on Mac. I originally created these test cases on my local Linux machine. Since I do not have good access to Windows to debug this, I set up a trial Azure account and ran the Parcel Windows node v12/v14 pipelines on this with some logging for the four proposed integration test cases. 

If I did this correctly, it appears that the 'browser', 'custom-target', and 'main' package.json test scenarios are not triggering the desired Parcel diagnostic error on Windows at all (despite correctly doing so on Linux), and the 'module' case on Windows triggers a different Parcel error - "esmodule output is not supported without scope hoisting" - instead of the intended error message that I get on Linux. Do you have any suggestions? Many thanks in advance for your thoughts.

File: packages/core/core/src/requests/TargetRequest.js [hunk 1]
```diff
@@ -91,6 +91,24 @@ async function run({input, api, options}: RunOpts) {
   );
   let targets = await targetResolver.resolve(input.packagePath, input.target);
 
+  for (const currTarget of targets) {
+    if (
+      currTarget.name !== 'default' &&
+      currTarget.distEntry != null &&
+      currTarget.distDir + '/' + currTarget.distEntry === input.filePath
```
Comment (DeMoorJasper) [inline packages/core/core/src/requests/TargetRequest.js:98]: not sure if this'll help in resolving the windows test case issue but you probably want to use path utils for this, something like `path.join(currTarget.distDir, currTarget.distEntry) === input.filePath` or even `path.join(currTarget.distDir, currTarget.distEntry) === path.normalize(input.filePath)`

Comment (JessieFrance) [reply packages/core/core/src/requests/TargetRequest.js:98]: @DeMoorJasper you're right. That code looks a lot cleaner. I'll use one of those if I'm able to debug this. Thank you.

**Comment by JessieFrance** [comment]
@DeMoorJasper your suggestion on using path utilities solved the previous issues my four test cases had on Windows. Thank you!

**Comment by devongovett** [comment]
@JessieFrance hope you don't mind me pushing to your branch. I added a bit more detail to the diagnostics, e.g. a code frame.

image
image

**Review by devongovett — APPROVED** [review]

**Pull request merged**